src/daemon/control.rs

```diff
@@ -479,7 +419,7 @@ pub fn onchain_txs_list_from_outpoints(
     revaultd: &RevaultD,
     bitcoind_tx: &Sender<BitcoindMessageOut>,
     outpoints: Option<Vec<OutPoint>>,
-) -> Result<Result<Vec<VaultOnchainTransactions>, RpcControlError>, ControlError> {
+) -> Result<Result<Vec<VaultOnchainTransactions>, RpcControlError>, RpcControlError> {
```

Comment:
...please no

I didn't like having nested errors before, but nested errors with the same error? :sob: Also it's not really clear why having them separated (I know why: because one is an `internal_error` and the other one is an `invalid_params`, but it's not obvious). I would merge the two errors (returning `Result<Vec<VaultOnchainTransactions, RpcControlError>` and then in `jsonrpc/api.rs` do something like
```
map_err(|e| match e {
... => invalid_params
... => internal_error
})?;
```
Maybe you can implement `From<RpcControlError> for jsonrpc_core::Result<>` so we don't need to map_err every time...

Reply:
Yeah, right :)

I don't think we should have a From but rather decide what kind of error to return on a case-by-case basis.

How about:
```diff
diff --git a/src/daemon/control.rs b/src/daemon/control.rs
index 11aeec9..4b291f8 100644
--- a/src/daemon/control.rs
+++ b/src/daemon/control.rs
@@ -8,7 +8,6 @@ use crate::{
         interface::{
             db_cancel_transaction, db_emer_transaction, db_signed_emer_txs, db_signed_unemer_txs,
             db_unvault_emer_transaction, db_unvault_transaction, db_vault_by_deposit, db_vaults,
-            db_vaults_min_status,
         },
         schema::DbVault,
         DatabaseError,
@@ -255,37 +254,44 @@ pub fn listvaults_from_db(
     })
 }
 
-/// List all the presigned transactions from these confirmed vaults.
-pub fn presigned_txs_list_from_outpoints(
-    revaultd: &RevaultD,
-    outpoints: Option<Vec<OutPoint>>,
-) -> Result<Result<Vec<VaultPresignedTransactions>, RpcControlError>, RpcControlError> {
-    let db_path = &revaultd.db_file();
-
-    // If they didn't provide us with a list of outpoints, catch'em all!
-    let db_vaults = if let Some(outpoints) = outpoints {
-        // FIXME: we can probably make this more efficient with some SQL magic
-        let mut vaults = Vec::with_capacity(outpoints.len());
-        for outpoint in outpoints.iter() {
-            if let Some(vault) = db_vault_by_deposit(db_path, &outpoint)? {
-                // If it's unconfirmed, the presigned transactions are not in db!
-                match vault.status {
-                    VaultStatus::Unconfirmed => {
-                        return Ok(Err(RpcControlError::InvalidStatus(
-                            vault.status,
-                            VaultStatus::Funded,
-                        )));
-                    }
-                    _ => vaults.push(vault),
+/// Get all vaults that are at least 'funded' from a list of outpoints.
+///
+/// # Errors
+/// If an outpoint does not refer to a known, at least 'funded' deposit.
+pub fn confirmed_vaults(
+    db_path: &std::path::PathBuf,
+    outpoints: Vec<OutPoint>,
+) -> Result<Vec<DbVault>, RpcControlError> {
+    let mut vaults = Vec::with_capacity(outpoints.len());
+
+    for outpoint in outpoints.iter() {
+        // Note: being smarter with SQL queries implies enabling the 'table' feature of rusqlite
+        // with a shit ton of dependencies.
+        if let Some(vault) = db_vault_by_deposit(db_path, &outpoint)? {
+            // If it's unconfirmed, the presigned transactions are not in db!
+            match vault.status {
+                VaultStatus::Unconfirmed => {
+                    return Err(RpcControlError::InvalidStatus(
+                        vault.status,
+                        VaultStatus::Funded,
+                    ));
                 }
-            } else {
-                return Ok(Err(RpcControlError::UnknownOutPoint(*outpoint)));
+                _ => vaults.push(vault),
             }
+        } else {
+            return Err(RpcControlError::UnknownOutPoint(*outpoint));
         }
-        vaults
-    } else {
-        db_vaults_min_status(db_path, VaultStatus::Funded)?
-    };
+    }
+
+    Ok(vaults)
+}
+
+/// List all the presigned transactions from these confirmed vaults.
+pub fn presigned_txs(
+    revaultd: &RevaultD,
+    db_vaults: Vec<DbVault>,
+) -> Result<Vec<VaultPresignedTransactions>, RpcControlError> {
+    let db_path = &revaultd.db_file();
 
     // For each presigned transaction, append it as well as its extracted version if it's final.
     let mut tx_list = Vec::with_capacity(db_vaults.len());
@@ -355,7 +361,7 @@ pub fn presigned_txs_list_from_outpoints(
         });
     }
 
-    Ok(Ok(tx_list))
+    Ok(tx_list)
 }
 
 /// List all the onchain transactions from these vaults.
diff --git a/src/daemon/jsonrpc/api.rs b/src/daemon/jsonrpc/api.rs
index ee1cdd2..1339339 100644
--- a/src/daemon/jsonrpc/api.rs
+++ b/src/daemon/jsonrpc/api.rs
@@ -5,10 +5,10 @@
 use crate::{
     control::{
         announce_spend_transaction, bitcoind_broadcast, check_revocation_signatures,
-        check_spend_signatures, check_unvault_signatures, fetch_cosigs_signatures,
-        finalized_emer_txs, listvaults_from_db, onchain_txs_list_from_outpoints,
-        presigned_txs_list_from_outpoints, share_rev_signatures, share_unvault_signatures,
-        ListSpendEntry, RpcUtils,
+        check_spend_signatures, check_unvault_signatures, confirmed_vaults,
+        fetch_cosigs_signatures, finalized_emer_txs, listvaults_from_db,
+        onchain_txs_list_from_outpoints, presigned_txs, share_rev_signatures,
+        share_unvault_signatures, ListSpendEntry, RpcUtils,
     },
     database::{
         actions::{
@@ -19,7 +19,7 @@ use crate::{
         interface::{
             db_cancel_transaction, db_emer_transaction, db_list_spends, db_spend_transaction,
             db_tip, db_unvault_emer_transaction, db_unvault_transaction, db_vault_by_deposit,
-            db_vault_by_unvault_txid, db_vaults_from_spend,
+            db_vault_by_unvault_txid, db_vaults_from_spend, db_vaults_min_status,
         },
     },
     jsonrpc::UserRole,
@@ -753,10 +753,19 @@ impl RpcApi for RpcImpl {
         meta: Self::Metadata,
         outpoints: Option<Vec<OutPoint>>,
     ) -> jsonrpc_core::Result<serde_json::Value> {
-        let vaults =
-            presigned_txs_list_from_outpoints(&meta.rpc_utils.revaultd.read().unwrap(), outpoints)
+        let revaultd = meta.rpc_utils.revaultd.read().unwrap();
+        let db_path = revaultd.db_file();
+
+        // If they didn't provide us with a list of outpoints, catch'em all!
+        let db_vaults = if let Some(outpoints) = outpoints {
+            confirmed_vaults(&db_path, outpoints)
+                .map_err(|e| JsonRpcError::invalid_params(e.to_string()))?
+        } else {
+            db_vaults_min_status(&db_path, VaultStatus::Funded)
                 .map_err(|e| internal_error!(e))?
-                .map_err(|e| JsonRpcError::invalid_params(e.to_string()))?;
+        };
+        let vaults = presigned_txs(&revaultd, db_vaults)
+            .map_err(|e| internal_error!(e))?;
 
         let vaults: Vec<serde_json::Value> = vaults
             .into_iter()
```

Reply:
Pushed as a new commit.